File: typescript/ldap_injection/userAuthentication.ts
import { Request, Response } from 'express';
import ldap from 'ldapjs';

const ldapClient = ldap.createClient({ url: 'ldap://shopist-directory.internal:389' });

interface LoginBody {
    username: string;
    password: string;
}

interface EmployeeQuery {
    employeeId: string;
    department?: string;
}

interface GroupQuery {
    username: string;
    groupName: string;
}

// VULN 1: String concatenation in ldapjs filter for authentication — attacker bypasses auth with *)(uid=*)
export function authenticateUser(req: Request, res: Response): void {
    const { username, password } = req.body as LoginBody;

    // Filter is built by string concatenation — attacker injects LDAP metacharacters to bypass authentication
    const searchFilter = '(&(objectClass=person)(uid=' + username + ')(userPassword=' + password + '))';
    const searchOptions: ldap.SearchOptions = {
        filter: searchFilter,
        scope: 'sub',
        attributes: ['uid', 'cn', 'mail', 'memberOf'],
    };

    ldapClient.search('ou=users,dc=shopist,dc=io', searchOptions, (err, result) => {
        const entries: ldap.SearchEntry[] = [];
        result.on('searchEntry', (entry: ldap.SearchEntry) => entries.push(entry));
        result.on('end', () => {
            if (entries.length > 0) {
                res.json({ authenticated: true, user: entries[0].object });
            } else {
                res.status(401).json({ error: 'Invalid credentials' });
            }
        });
    });
}

// VULN 2: Template literal in employee lookup filter — injection via employeeId parameter
export function lookupEmployee(req: Request, res: Response): void {
    const { employeeId, department } = req.query as unknown as EmployeeQuery;

    // Template literal builds LDAP filter with unescaped user input — attacker leaks all directory entries
    const searchFilter = `(&(objectClass=employee)(employeeNumber=${employeeId})(department=${department}))`;
    const searchOptions: ldap.SearchOptions = {
        filter: searchFilter,
        scope: 'sub',
        attributes: ['cn', 'mail', 'telephoneNumber', 'title'],
    };

    ldapClient.search('ou=employees,dc=shopist,dc=io', searchOptions, (err, result) => {
        const entries: ldap.SearchEntry[] = [];
        result.on('searchEntry', (entry: ldap.SearchEntry) => entries.push(entry));
        result.on('end', () => res.json({ employees: entries.map(e => e.object) }));
    });
}

// VULN 3: String concatenation in group membership filter — attacker enumerates all LDAP groups
export function checkGroupMembership(req: Request, res: Response): void {
    const { username, groupName } = req.body as GroupQuery;

    // Both username and groupName are concatenated directly — LDAP injection grants unauthorized group access
    const memberFilter = '(&(objectClass=groupOfNames)(cn=' + groupName + ')(member=uid=' + username + ',ou=users,dc=shopist,dc=io))';
    const searchOptions: ldap.SearchOptions = {
        filter: memberFilter,
        scope: 'sub',
        attributes: ['cn', 'description'],
    };

    ldapClient.search('ou=groups,dc=shopist,dc=io', searchOptions, (err, result) => {
        const groups: ldap.SearchEntry[] = [];
        result.on('searchEntry', (entry: ldap.SearchEntry) => groups.push(entry));
        result.on('end', () => {
            res.json({ isMember: groups.length > 0, groups: groups.map(g => g.object) });
        });
    });
}

File: dotnet/ldap_injection/UserAuthentication.cs
using System.DirectoryServices;

public class UserAuthentication
{
    private const string LdapPath = "LDAP://dc.shopist.internal/DC=shopist,DC=com";

    // VULN 1: String concatenation used to build DirectorySearcher.Filter for LDAP authentication
    public bool AuthenticateUser(string username, string password)
    {
        using var entry = new DirectoryEntry(LdapPath);
        using var searcher = new DirectorySearcher(entry);
        searcher.Filter = "(&(objectClass=user)(sAMAccountName=" + username + ")(userPassword=" + password + "))";
        SearchResult result = searcher.FindOne();
        return result != null;
    }

    // VULN 2: String.Format used to build DirectorySearcher.Filter for employee directory lookup
    public SearchResult LookupEmployee(string department, string employeeName)
    {
        using var entry = new DirectoryEntry(LdapPath);
        using var searcher = new DirectorySearcher(entry);
        searcher.Filter = string.Format("(&(objectClass=person)(department={0})(cn={1}))", department, employeeName);
        return searcher.FindOne();
    }

    // VULN 3: String concatenation used to build DirectorySearcher.Filter for group membership check
    public bool IsUserInGroup(string username, string groupName)
    {
        using var entry = new DirectoryEntry(LdapPath);
        using var searcher = new DirectorySearcher(entry);
        searcher.Filter = "(&(objectClass=user)(sAMAccountName=" + username + ")(memberOf=CN=" + groupName + ",OU=Groups,DC=shopist,DC=com))";
        SearchResult result = searcher.FindOne();
        return result != null;
    }
}
